src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
+
+#if DEBUG
+            if (Logging.On) Logging.Enter(Logging.Http, this, "StartRequest", state.RequestMessage);
+#endif
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
                 return;
             }
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+            var cancellationTokenRegistration = state.CancellationToken.Register(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.TrySetCanceled(rcs.CancellationToken);
+                },
+                state);
 
             try
             {
-                requestHandle = CreateRequestHandle(state);
+                // Prepare context objects
+                requestStateHandle = GCHandle.Alloc(state);
+                responseMessage = new CurlResponseMessage(state.RequestMessage);
+                state.ResponseMessage = responseMessage;
+                requestHandle = CreateRequestHandle(state, requestStateHandle, out headerPtr);
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                int result = Interop.libcurl.curl_easy_perform(requestHandle);
-                if (CURLcode.CURLE_OK != result)
+                if (state.CancellationToken.IsCancellationRequested)
                 {
-                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
                 }
 
-                // TODO: Handle requests with body
+                if (state.RequestMessage.Content != null)
+                {
+#if DEBUG
+                    Logging.PrintInfo(Logging.Http, this, "Request content for " + Logging.GetObjectLogHash(responseMessage.ContentStream) + " : " + state.RequestMessage.Content.ReadAsStringAsync().Result);
+#endif
+
+                    Stream requestContentStream =
+                        await state.RequestMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                    if (state.CancellationToken.IsCancellationRequested)
+                    {
+                        state.TrySetCanceled(state.CancellationToken);
+                        return;
+                    }
+                    state.RequestContentStream = requestContentStream;
+                }
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                requestHandle.DangerousAddRef(ref referenceTaken);
+                var performTask = Task.Run(() =>
+                    {
+                        if (state.CancellationToken.IsCancellationRequested)
```

Comment:
We're checking IsCancellationRequested manually at a whole bunch of places.  Do we need to be?  IsCancellationRequested is cheap, so this shouldn't hurt performance, but it does impact readability/maintainability of the code, and I'm not sure there's much benefit to checking it explicitly like this every step of the way.